tools/migration/internal/repo_fs_wrangler.go

```diff
@@ -1,68 +1,88 @@
 package internal
 
 import (
+	"errors"
+	"fmt"
 	"os"
 	"strings"
+
+	rcopy "github.com/otiai10/copy"
 )
 
-// RepoFSWrangler manages filesystem operations and figures out what the correct paths
-// are for everything.
-type RepoFSWrangler struct {
-	oldRepoPath, newRepoPath string
-}
+// CloneRepo copies the old repo to the new repo dir with Read/Write access.
+//   Returns an error if the directory exists.
+func CloneRepo(oldRepoPath string) (string, error) {
+	realRepoPath, err := os.Readlink(oldRepoPath)
+	if err != nil {
+		return "", fmt.Errorf("old-repo must be a symbolic link: %s", err)
+	}
 
-// NewRepoFSWrangler takes options for old and new repo paths, figures out
-// what the correct paths should be, and creates a new RepoFSWrangler with the
-// correct paths.
-func NewRepoFSWrangler(oldRepoOpt, newRepoPrefixOpt, oldVersion, newVersion string) *RepoFSWrangler {
-	return &RepoFSWrangler{
-		newRepoPath: getNewRepoPath(oldRepoOpt, newRepoPrefixOpt, oldVersion, newVersion),
-		oldRepoPath: oldRepoOpt,
+	newRepoPath, err := getNewRepoPath(oldRepoPath, "")
+	if err != nil {
+		return "", err
 	}
-}
 
-// GetOldRepo returns the old repo dir, opened as read-only.
-func (rmh *RepoFSWrangler) GetOldRepo() (*os.File, error) {
-	return os.Open(rmh.oldRepoPath)
+	if err := rcopy.Copy(realRepoPath, newRepoPath); err != nil {
+		return "", err
+	}
+	if err := os.Chmod(newRepoPath, os.ModeDir|0744); err != nil {
+		return "", err
+	}
+	return newRepoPath, nil
 }
 
-// GetOldRepoPath returns the path to the existing repo
-func (rmh *RepoFSWrangler) GetOldRepoPath() string {
-	return rmh.oldRepoPath
-}
+// InstallNewRepo archives the old repo, and symlinks the new repo in its place.
+// returns the new path to the old repo and any error.
+func InstallNewRepo(oldRepoLink, newRepoPath string) (string, error) {
+	realRepoPath, err := os.Readlink(oldRepoLink)
+	if err != nil {
+		return "", err
+	}
 
-// GetNewRepoPath returns the path to the new repo. It makes no guarantees about
-//   whether the directory exists.
-func (rmh *RepoFSWrangler) GetNewRepoPath() string {
-	return rmh.newRepoPath
-}
+	archivedRepoPath := strings.Join([]string{realRepoPath, "archived", NowString()}, "-")
 
-// MakeNewRepo creates the new repo dir and returns it with Read/Write access.
-//   This returns an error if the directory exists.
-func (rmh *RepoFSWrangler) MakeNewRepo() (*os.File, error) {
-	if err := os.Mkdir(rmh.newRepoPath, os.ModeDir|0744); err != nil {
-		return nil, err
+	if err := os.Rename(realRepoPath, archivedRepoPath); err != nil {
+		return archivedRepoPath, err
+	}
```

Comment:
Blocking: this should happen *after* the symlink is switched, else subsequent failure leaves the link broken.

Reply:
I assume you mean the rename and not the link removal, so if we're not renaming then it's not an issue.